exporters/otlp/src/otlp_http_client.cc

```diff
@@ -563,31 +644,200 @@ void ConvertListFieldToJson(nlohmann::json &value,
+    std::size_t max_running_requests) noexcept
+{
+  auto session = createSession(message, std::move(result_callback));
+  if (opentelemetry::nostd::holds_alternative<sdk::common::ExportResult>(session))
+  {
+    sdk::common::ExportResult result =
+        opentelemetry::nostd::get<sdk::common::ExportResult>(session);
+    if (result_callback)
     {
-      std::cerr << error_message << std::endl;
+      result_callback(result);
     }
-    OTEL_INTERNAL_LOG_ERROR(error_message);
+    return result;
+  }
+
+  addSession(std::move(opentelemetry::nostd::get<HttpSessionData>(session)));
 
+  // Wait for the response to be received
+  if (options_.console_debug)
+  {
+    OTEL_INTERNAL_LOG_DEBUG(
+        "[OTLP HTTP Client] DEBUG: Waiting for response from "
+        << options_.url << " (timeout = "
+        << std::chrono::duration_cast<std::chrono::milliseconds>(options_.timeout).count()
+        << " milliseconds)");
+  }
+
+  // Wait for any session to finish if there are to many sessions
+  std::unique_lock<std::mutex> lock(session_waker_lock_);
+  bool wait_successful =
+      session_waker_.wait_for(lock, options_.timeout, [this, max_running_requests] {
+        std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+        return running_sessions_.size() <= max_running_requests;
+      });
+
+  cleanupGCSessions();
+
+  if (!wait_successful)
+  {
     return opentelemetry::sdk::common::ExportResult::kFailure;
   }
 
+  return opentelemetry::sdk::common::ExportResult::kSuccess;
+}
+
+bool OtlpHttpClient::Shutdown(std::chrono::microseconds timeout) noexcept
+{
+  {
+    std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+    is_shutdown_ = true;
+
+    // Shutdown the session manager
+    http_client_->CancelAllSessions();
+    http_client_->FinishAllSessions();
+  }
+
+  // ASAN will report chrono: runtime error: signed integer overflow: A + B cannot be represented
+  //   in type 'long int' here. So we reset timeout to meet signed long int limit here.
+  timeout = opentelemetry::common::DurationUtil::AdjustWaitForTimeout(
+      timeout, std::chrono::microseconds::zero());
+
+  // Wait for all the sessions to finish
+  std::unique_lock<std::mutex> lock(session_waker_lock_);
+  if (timeout <= std::chrono::microseconds::zero())
+  {
+    while (true)
+    {
+      {
+        std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+        if (running_sessions_.empty())
+        {
+          break;
+        }
+      }
+      // When changes of running_sessions_ and notify_one/notify_all happen between predicate
+      // checking and waiting, we should not wait forever.
+      session_waker_.wait_for(lock, options_.timeout);
+    }
+  }
+  else
+  {
+    session_waker_.wait_for(lock, timeout, [this] {
+      std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+      return running_sessions_.empty();
+    });
+  }
+
+  while (cleanupGCSessions())
```

Comment:
There is multiple `while` loops in this `Shutdown` method - are we sure they won't ever hang the shutdown?